src/serializer/nodeSerializers.ts

```diff
@@ -0,0 +1,402 @@
+  }
+
+  private indent!: IndentationManager
+
+  // eslint-disable-next-line complexity
+  private serializeNode(node: ExprNode): string {
+    switch (node.type) {
+      case 'Value':
+        return this.serializeValue(node)
+      case 'Everything':
+        return '*'
+      case 'This':
+        return '@'
+      case 'Parent':
+        return '^'.repeat(node.n)
+      case 'Parameter':
+        return `$${node.name}`
+      case 'AccessAttribute':
+        return this.serializeAccessAttribute(node)
+      case 'AccessElement':
+        return this.serializeAccessElement(node)
+      case 'Array':
+        return this.serializeArray(node)
+      case 'ArrayCoerce':
+        return this.serializeArrayCoerce(node)
+      case 'Object':
+        return this.serializeObject(node)
+      case 'OpCall':
+        return this.serializeOpCall(node)
+      case 'And':
+        return this.serializeBinaryOp(node.left, '&&', node.right)
+      case 'Or':
+        return this.serializeBinaryOp(node.left, '||', node.right)
+      case 'Not':
+        return this.serializeUnaryOp('!', node.base)
+      case 'Neg':
+        return this.serializeUnaryOp('-', node.base)
+      case 'Pos':
+        return this.serializeUnaryOp('+', node.base)
+      case 'Group':
+        return this.serializeGroup(node)
+      case 'FuncCall':
+        return this.serializeFuncCall(node)
+      case 'PipeFuncCall':
+        return this.serializePipeFuncCall(node)
+      case 'Deref':
+        return this.serializeDeref(node)
+      case 'Filter':
+        return this.serializeFilter(node)
+      case 'Projection':
+        return this.serializeProjection(node)
+      case 'Slice':
+        return this.serializeSlice(node)
+      case 'InRange':
+        return this.serializeInRange(node)
+      case 'Select':
+        return this.serializeSelect(node)
+      case 'Asc':
+        return this.serializeAsc(node)
+      case 'Desc':
+        return this.serializeDesc(node)
+      case 'Tuple':
+        return this.serializeTuple(node)
+      case 'Map':
+        return this.serializeMap(node)
+      case 'FlatMap':
+        return this.serializeFlatMap(node)
+      case 'Context':
+        return this.serializeContext(node)
+      case 'Selector':
+        return '<selector>'
+      default:
+        // @ts-expect-error handle all cases
+        throw new Error(`Unknown node type: ${node.type}`)
+    }
+  }
+
+  private serializeValue(node: ValueNode): string {
+    const value = node.value
+    if (typeof value === 'string') {
+      // https://spec.groq.dev/GROQ-1.revision3/#sec-String
+      const backspace = String.fromCharCode(8) // U+0008 backspace
+      return `"${value
+        .replace(/\\/g, '\\\\')
+        .replace(/"/g, '\\"')
+        .replace(new RegExp(backspace, 'g'), '\\b')
+        .replace(/\f/g, '\\f')
+        .replace(/\n/g, '\\n')
+        .replace(/\r/g, '\\r')
+        .replace(/\t/g, '\\t')}"`
+    }
+    return String(value)
```

Comment:
`JSON.stringify` should return a compatible string without regex foo:
```suggestion
    return JSON.stringify(node.value)
```